ast/json_test.go

```diff
@@ -157,6 +157,308 @@ func TestJSONMarshal(t *testing.T) {
+				Properties: []*ast.PropertyType{
+					{
+						BaseNode: ast.BaseNode{},
+						Name: &ast.Identifier{
+							BaseNode: ast.BaseNode{},
+							Name:     "A",
+						},
+						Ty: &ast.NamedType{
+							BaseNode: ast.BaseNode{},
+							ID: &ast.Identifier{
+								BaseNode: ast.BaseNode{},
+								Name:     "int",
+							},
+						},
+					},
+				},
+				Tvar: nil,
+			},
+			// {"type":"RecordType","properties":[{"name":{"name":"A"},"monotype":{"type":"NamedType","name":{"name":"int"}}}]}"#
+			want: `{"type":"RecordType","properties":[{"type":"PropertyType","name":{"type":"Identifier","name":"A"},"ty":{"type":"NamedType","name":{"type":"Identifier","name":"int"}}}],"tvar":null}`,
+		},
+		{
+			name: "FunctionType_NoParams",
+			node: &ast.FunctionType{
+				BaseNode:   ast.BaseNode{},
+				Parameters: []*ast.ParameterType{},
+				Return: &ast.NamedType{
+					BaseNode: ast.BaseNode{},
+					ID: &ast.Identifier{
+						BaseNode: ast.BaseNode{},
+						Name:     "int",
+					},
+				},
+			},
+			want: `{"type":"FunctionType","parameters":[],"return":{"type":"NamedType","name":{"type":"Identifier","name":"int"}}}`,
+		},
+		{
+			name: "FunctionType_Required",
+			node: &ast.FunctionType{
+				BaseNode: ast.BaseNode{},
+				Parameters: []*ast.ParameterType{
+					{
+						BaseNode: ast.BaseNode{},
+						Name: &ast.Identifier{
+							BaseNode: ast.BaseNode{},
+							Name:     "B",
+						},
+						Ty: &ast.NamedType{
+							BaseNode: ast.BaseNode{},
+							ID: &ast.Identifier{
+								BaseNode: ast.BaseNode{},
+								Name:     "string",
+							},
+						},
+						Kind: 0},
+				},
+				Return: &ast.NamedType{
+					BaseNode: ast.BaseNode{},
+					ID: &ast.Identifier{
+						BaseNode: ast.BaseNode{},
+						Name:     "uint",
+					},
+				},
+			},
+			want: `{"type":"FunctionType","parameters":[{"type":"ParameterType","name":{"type":"Identifier","name":"B"},"ty":{"type":"NamedType","name":{"type":"Identifier","name":"string"}},"kind":0}],"return":{"type":"NamedType","name":{"type":"Identifier","name":"uint"}}}`,
+		},
+		{
+			name: "FunctionType_Optional",
+			node: &ast.FunctionType{
+				BaseNode: ast.BaseNode{},
+				Parameters: []*ast.ParameterType{
+					{
+						BaseNode: ast.BaseNode{},
+						Name: &ast.Identifier{
+							BaseNode: ast.BaseNode{},
+							Name:     "A",
+						},
+						Ty: &ast.NamedType{
+							BaseNode: ast.BaseNode{},
+							ID: &ast.Identifier{
+								BaseNode: ast.BaseNode{},
+								Name:     "int",
+							},
+						},
+						Kind: 1},
+				},
+				Return: &ast.NamedType{
+					BaseNode: ast.BaseNode{},
+					ID: &ast.Identifier{
+						BaseNode: ast.BaseNode{},
+						Name:     "int",
+					},
+				},
+			},
+			want: `{"type":"FunctionType","parameters":[{"type":"ParameterType","name":{"type":"Identifier","name":"A"},"ty":{"type":"NamedType","name":{"type":"Identifier","name":"int"}},"kind":1}],"return":{"type":"NamedType","name":{"type":"Identifier","name":"int"}}}`,
```

Comment:
The parameter type json doesn't quite look like the Rust version. Can you add the rust version as a comment right above this line like you did with some of the others?

Reply:
I pushed the code with the rust commented above, since the format for Go is the "kinds" and not the keyword "optional" the syntax is different. Should I change that?